### PR TITLE
Improve specification of selectors

### PR DESCRIPTION
**Pull request by judofyr**
- Fully describe the syntax
- Fully describe the search/evaluation semantic
- Fix usage in changedAny/changedOnly

I _think_ this is correct now. I think I want to take a second look later to double-check everything 😅

### CODE REVIEW

**Comment by judofyr** [comment]
(Fixed formatting and force-pushed)

File: spec/Section 3 -- Execution.md [hunk 1]
```diff
@@ -115,22 +115,59 @@ A boosted predicate simply adds the boost value to the score if the predicate ma
...
+          - Let {matched} be the result of {Evaluate(expr, nestedScope)}.
+          - If {matched} is {true}:
+            - Let {nestedKeyPath} be the result of combining {keyPath} with the array index.
+            - Append {nestedKeyPath} to {result}.
+    - If {selector} ends with a {SelectorGroup}:
+      - Let {inner} be that selector.
+      - Let {innerResult} be the result of {EvaluateSelector(inner, innerValue, scope)}.
+      - For each {nestedKeyPath} in {innerResult}:
+        - Let {combinedKeyPath} be the result of combining {keyPath} with {nestedKeyPath}.
+        - Append {keyPath} to {result}.
```
Comment (sgulseth) [inline spec/Section 3 -- Execution.md:163]: Combined key path? 🤔

File: spec/Section 3 -- Execution.md [hunk 2]
```diff
@@ -115,22 +115,59 @@ A boosted predicate simply adds the boost value to the score if the predicate ma
...
+    - If {selector} ends with a {SelectorTuple}:
+      - For each selector {inner} in the tuple:
+        - Let {innerResult} be the result of {EvaluateSelector(inner, innerValue, scope)}.
+        - For each {nestedKeyPath} in {innerResult}:
+          - Let {combinedKeyPath} be the result of combining {keyPath} with {nestedKeyPath}.
+          - Append {keyPath} to {result}.
```
Comment (sgulseth) [inline spec/Section 3 -- Execution.md:169]: Same as above

File: spec/Section 3 -- Execution.md [hunk 3]
```diff
@@ -115,22 +115,59 @@ A boosted predicate simply adds the boost value to the score if the predicate ma
...
+  - If {value} is an object which has the given attribute:
+    - Let {keyPath} be a new key path consisting of the attribute name.
+    - Append {keyPath} to {result}.
+- If {selector} starts with a {Selector}:
+  - Let {baseSelector} be the selector.
+  - Let {base} be the result of {EvaluateSelector(baseSelector, value, scope)}.
+  - For each {keyPath} in {base}:
+    - Let {innerValue} be the value at {keyPath} in {value}.
+    - If {selector} ends with a {ArrayPostfix}:
+      - If {innerValue} is an array:
```
Comment (sgulseth) [inline spec/Section 3 -- Execution.md:142]: What if it's not an array(or other types assumed below), should the spec cover errors or let that be up to the impl?

**Review by sgulseth — CHANGES_REQUESTED** [review]

File: spec/Section 3 -- Execution.md (hunk 3 above)
Comment (judofyr) [reply spec/Section 3 -- Execution.md:142]: It shouldn't cause any error at all. It should just _not_ match up.

File: spec/Section 3 -- Execution.md (hunk 1 above)
Comment (judofyr) [reply spec/Section 3 -- Execution.md]: Yes. Fixed in force-push.

File: spec/Section 3 -- Execution.md (hunk 2 above)
Comment (judofyr) [reply spec/Section 3 -- Execution.md]: Yes. Fixed in force-push.

File: spec/Section 3 -- Execution.md (hunk 3 above)
Comment (sgulseth) [reply spec/Section 3 -- Execution.md:142]: ✨

**Review by sgulseth — APPROVED** [review]

**Review by sgulseth — APPROVED** [review]

File: spec/Section 3 -- Execution.md [hunk 4]
```diff
@@ -115,22 +115,59 @@ A boosted predicate simply adds the boost value to the score if the predicate ma
...
+          - Let {nestedKeyPath} be the result of combining {keyPath} with the array index.
+          - Append {nestedKeyPath} to {result}.
+    - If {selector} ends with a {AttributeAccess}:
+      - If {innerValue} is an object which has the given attribute:
+        - Let {nestedKeyPath} be the result of combining {keyPath} with the attribute name.
+        - Append {nestedKeyPath} to {result}.
+    - If {selector} ends with a {Filter}:
+      - If {innerValue} is an array:
+        - For each {item} of {innerValue}:
+          - Let {nestedScope} be the result of {NewNestedScope(value, scope)}.
```
Comment (codebymatt) [inline spec/Section 3 -- Execution.md:153]: What is `NewNestedScope`? Is it defined elsewhere in the spec?

**Review by codebymatt — APPROVED** [review]
Nice work! I have a suggestion and a question, but neither are worth blocking.

File: spec/Section 3 -- Execution.md (hunk 4 above)
Comment (judofyr) [reply spec/Section 3 -- Execution.md:153]: Yup, it's here: https://sanity-io.github.io/GROQ/draft/#NewNestedScope()

File: spec/Section 13 -- Extensions.md [hunk 5]
```diff
@@ -190,10 +190,9 @@ diff::changedAny(args, scope):
 - Let {selector} be the third element of {args}.
 - Let {before} be the result of {Evaluate(lhs, scope)}.
 - Let {after} be the result of {Evaluate(rhs, scope)}.
-- Let {evaluatedSelector} be the result of {EvaluateSelector(selector, scope)}
-- Let {search} be the result of {SelectorSearch(evaluatedSelector, before, scope)}
-- Let {diff} be the result of comparative difference between {before} and {after}
-- If {diff} overlaps with {search}:
+- Let {beforeKeyPaths} be the result of {EvaluateSelector(selector, before, scope)}.
```
[the remark this answers is not in the document]
Comment (judofyr) [reply spec/Section 13 -- Extensions.md]: Done!

**Pull request merged**